grabbit/core.py

```diff
@@ -1,26 +1,164 @@
 import json
+import logging
 import os
 import re
+import shutil
+import sys
 from collections import defaultdict, OrderedDict, namedtuple
 from grabbit.external import six, inflect
-from grabbit.utils import natural_sort
-from os.path import join, basename, dirname, abspath, split
+from grabbit.utils import natural_sort, splitext
+from os.path import (join, basename, dirname, abspath, split, exists, islink,
+                     isabs, isdir)
 from functools import partial
+from six import string_types
 
 
 __all__ = ['File', 'Entity', 'Layout']
 
 
 class File(object):
 
-    def __init__(self, filename):
+    def __init__(self, filename, path_patterns=None):
         """
         Represents a single file.
         """
         self.path = filename
         self.filename = basename(self.path)
         self.dirname = dirname(self.path)
         self.entities = {}
+        self.path_patterns = path_patterns
+
+    def replace_entities(self, pattern):
+        new_path = pattern
+        ents = re.findall('\{(.*?)\}', pattern)
+        ents_matched = True
+        for ent in ents:
```

Comment:
I could be wrong, but I think you may be able to do something like `if not (set(ents) - set(self.entities.keys()))` to avoid the for-loop. (You'd still need to replace them all after that, but I think the code is clearer).